spec/jobs/send_gpo_code_expiration_notices_job_spec.rb

```diff
@@ -0,0 +1,242 @@
+
+  let!(:user_with_expired_code_who_should_be_notified) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: expired_and_notifiable_timestamp)
+  end
+
+  let!(:user_with_code_thats_not_expired_enough) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: expired_but_not_yet_notifiable_timestamp)
+  end
+
+  let!(:user_with_two_expired_and_notifiable_codes) do
+    create(
+      :user,
+      :with_pending_gpo_profile,
+      code_sent_at: expired_and_notifiable_timestamp,
+    ).tap do |user|
+      profile = user.gpo_verification_pending_profile
+      create(
+        :gpo_confirmation_code,
+        profile: profile,
+        code_sent_at: expired_and_notifiable_timestamp - 1.second,
+      )
+    end
+  end
+
+  let!(:user_who_already_got_an_expiration_notice) do
+    create(
+      :user,
+      :with_pending_gpo_profile,
+      code_sent_at: expired_and_notifiable_timestamp,
+    ).tap do |user|
+      code = user.
+        gpo_verification_pending_profile.
+        gpo_confirmation_codes.first
+
+      code.update(expiration_notice_sent_at: Time.zone.now)
+    end
+  end
+
+  let!(:user_who_completed_gpo) do
+    create(
+      :user,
+      :with_pending_gpo_profile,
+      code_sent_at: expired_and_notifiable_timestamp,
+    ).tap do |user|
+      profile = user.gpo_verification_pending_profile
+      profile.remove_gpo_deactivation_reason
+      profile.activate
+    end
+  end
+
+  let(:deactivation_reasons) do
+    %i[
+      password_reset
+      encryption_error
+      verification_cancelled
+      gpo_verification_pending_NO_LONGER_USED
+      in_person_verification_pending_NO_LONGER_USED
+    ]
+  end
+
+  let!(:users_with_profiles_with_deactivation_reasons) do
+    deactivation_reasons.map do |reason|
+      create(
+        :user,
+        :with_pending_gpo_profile,
+        code_sent_at: expired_and_notifiable_timestamp,
+      ).tap do |user|
+        profile = user.gpo_verification_pending_profile
+        profile.deactivate(reason)
+      end
+    end
+  end
+
+  let!(:user_who_started_gpo_but_then_verified_a_different_way) do
+    create(
+      :user,
+      :with_pending_gpo_profile,
+      code_sent_at: expired_and_notifiable_timestamp,
+    ).tap do |user|
+      create(:profile, :active, :verified, :with_pii, user: user)
+    end
+  end
+
+  let!(:user_who_has_an_unexpired_code) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: not_expired_timestamp)
+  end
+
+  let!(:user_with_code_that_expired_too_long_ago) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: too_expired_to_notify_timestamp)
+  end
+
+  before do
+    allow(IdentityConfig.store).to receive(:usps_confirmation_max_days).
+      and_return(usps_confirmation_max_days)
+  end
+
+  describe '#calculate_notification_window_bounds' do
+    it 'finds the right bounds' do
+      bounds = job.calculate_notification_window_bounds(
+        as_of: Time.zone.parse('2023-11-25 13:14:15'),
```

Comment:
Did you intentionally include the daylight -> standard time change in your calculation? Nice.